cloudsync/smartsync.py

```diff
@@ -447,3 +469,125 @@ def smart_rename(self, side, oid, new_path) -> str:
+        self.discarded_paths: set = set()
+        self.csync = csync
+
+    def handle_notification(self, notification: Notification):
+        """ implementation of callback that logs when files are discarded or skipped by SmartSync """
+        n = notification
+        if n.ntype not in (NotificationType.SYNC_DISCARDED, NotificationType.SYNC_SMART_UNSYNCED):
+            return  # not interested in non-sync related events
+
+        if n.ntype == NotificationType.SYNC_SMART_UNSYNCED and n.source == SourceEnum.LOCAL:
+            return  # only interested in REMOTE events, because smartsync operates primarily remotely
+
+        if n.ntype == NotificationType.SYNC_SMART_UNSYNCED:
+            self.skipped_paths.add(n.path)
+        elif n.ntype == NotificationType.SYNC_DISCARDED:
+            self.discarded_paths.add(n.path)
+
+    @staticmethod
+    def _path_in(path, paths, provider):
+        """ Checks if path is in path_dict, the keys of which are remote paths """
+        for candidate in paths:
+            if provider.paths_match(path, candidate):
+                return True
+        return False
+
+    def _is_synced(self, side, path, hash_str):
+        info: OInfo = self.csync.providers[side].info_path(path)
+        return info and (not hash_str or info.hash == hash_str)
+
+    def clear_sync_state(self):
+        """ Resets the log of synced/skipped paths """
+        self.skipped_paths = set()
+        self.discarded_paths = set()
+
+    def check_sync_state(  # pylint: disable=too-many-branches
+            self,
+            *,
+            remote_paths: Optional[Union[List[str], List[Tuple[str, str]]]] = None,  # tuple is (path, hash)
+            local_paths: Optional[Union[List[str], List[Tuple[str, str]]]] = None,
+            skipped_paths: Optional[Union[List[str], List[Tuple[str, str]]]] = None,
+            discarded_paths: Optional[Union[List[str], List[Tuple[str, int]]]] = None,  # tuple is (path, side)
+            quiet=False
+    ):
+        """ Returns True if synced_paths have synced and skipped_paths have explicitly been skipped """
+        if not (remote_paths or local_paths or skipped_paths or discarded_paths):
+            raise ValueError("Specify remote_paths or local_paths or skipped_paths or discarded_paths")
+
+        retval = True
+        for path in remote_paths or []:
+            h = None
+            if isinstance(path, tuple):
+                path, h = path
+            if not self._is_synced(REMOTE, path, h):
+                if not quiet:
+                    log.error("%s not synced remotely", path)
+                retval = False
+
+        for path in local_paths or []:
+            h = None
+            if isinstance(path, tuple):
+                path, h = path
+            if not self._is_synced(LOCAL, path, h):
+                if not quiet:
+                    log.error("%s not synced locally", path)
+                    self._is_synced(LOCAL, path, h)
```

Comment:
this line appears to have to effect

Reply:
Originally for debugging, forgot to remove. good catch